### PR TITLE
Added iteration timing to history file

### LINKED ISSUE #147: Add iteration timing
## Type of issue
What types of issue is it?
_Select the appropriate type(s) that describe this issue_

- New feature (non-breaking change which adds functionality)

## Description
I propose that we add a new key called `time` under each `callCounter`, which stores the wall time elapsed in seconds since the beginning of the optimization. This would allow an optimization history to be plotted against wall time, in addition to the callCounter available.

### PR DESCRIPTION
**Pull request by nwu63**
## Purpose
This was a popular feature request. Closes #147. The history file will now have `time` embedded in each iteration, which is helpful for plotting values against time instead of just iterations. Unfortunately the implementation is a bit crude since the same boilerplate code is in each optimizer wrapper---see #220. Nevertheless, this works and I've added some tests also.

I also added a basic`tearDown` implementation to clean up some of the generated test files, but this is incomplete.

## Type of change

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

## Testing
I added some tests.

## Checklist

- [x] I have run `flake8` and `black` to make sure the code adheres to PEP-8 and is consistently formatted
- [x] I have run unit and regression tests which pass locally with my changes
- [x] I have added new tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation

### CODE REVIEW

**Comment by ArshSaja** [comment]
Some style checks are failing. Is it okay to move on ?

**Comment by nwu63** [comment]
I've fixed all the violations

File: tests/testing_utils.py [hunk 1]
```diff
@@ -432,3 +443,13 @@ def optimize_with_hotstart(self, tol, optOptions=None, x0=None):
         # we should have zero actual function/gradient evaluations
         self.assertEqual(self.nf, 0)
         self.assertEqual(self.ng, 0)
+
+    def tearDown(self):
+        # remove history file
```
Comment (marcomangano) [inline tests/testing_utils.py:448]: What is missing here? It looks like it is covering all the generated files

**Review by marcomangano — COMMENTED** [review]
This is nice! ...but, given that we already have `metadata['optTime']`, wouldn't it be more useful to store the single iteration time rather than the cumulative time at iteration n? I know it's an easy post-production step and might mess with the test you set up, but I feel the info would me easier to interpret, especially if using OptView or `dictview`

**Comment by nwu63** [comment]
> This is nice! ...but, given that we already have `metadata['optTime']`, wouldn't it be more useful to store the single iteration time rather than the cumulative time at iteration n? I know it's an easy post-production step and might mess with the test you set up, but I feel the info would me easier to interpret, especially if using OptView or `dictview`

I think from a postprocessing perspective cumulative time makes more sense, as you would usually want to plot DV etc. against cumulative time to replace the iteration counters. What would be the use case for single iteration time?

File: tests/testing_utils.py (hunk 1 above)
Comment (nwu63) [reply tests/testing_utils.py:448]: Didn't get rid of all the files for me when I tried this. Don't have time to fix this though.

**Comment by marcomangano** [comment]
> > This is nice! ...but, given that we already have `metadata['optTime']`, wouldn't it be more useful to store the single iteration time rather than the cumulative time at iteration n? I know it's an easy post-production step and might mess with the test you set up, but I feel the info would me easier to interpret, especially if using OptView or `dictview`
> 
> I think from a postprocessing perspective cumulative time makes more sense, as you would usually want to plot DV etc. against cumulative time to replace the iteration counters. What would be the use case for single iteration time?

Oh I see, makes sense! From my perspective, because my cases are annoyingly long, it could be useful to monitor iter-time vs iteration as well. But you are right, from a general user perspective the cumulative time might be more relevant

**Review by ArshSaja — APPROVED** [review]
Thanks Neil..

**Review by marcomangano — APPROVED** [review]

**Pull request merged**